### PR TITLE
feat: configure database connection retries

### PR DESCRIPTION
**Pull request by grace0925**
Signed-off-by: Grace Liu <grace.liu@securekey.com>

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -26,7 +26,7 @@ cms
 .idea
 .DS_Store
 
-coverage.txt
+coverage.out
```
Comment (DRK3) [inline .gitignore:29]: ?

File: cmd/edv-rest/startcmd/start.go [hunk 2]
```diff
@@ -59,6 +61,14 @@ const (
 		" creating or accessing underlying databases." +
 		" Alternatively, this can be set with the following environment variable: " + databasePrefixEnvKey
 
+	databaseTimeoutFlagName      = "database-timeout"
+	databaseTimeoutEnvKey        = "EDV_DATABASE_TIMEOUT"
+	databaseTimeoutFlagShorthand = "o"
+	databaseTimeoutFlagUsage     = "Total time in seconds to wait until the datasource is available before giving up." +
```
Comment (DRK3) [inline cmd/edv-rest/startcmd/start.go:67]: I think you mean "database" not "datasource"

File: cmd/edv-rest/startcmd/start.go [hunk 3]
```diff
@@ -85,6 +95,8 @@ const (
 	tlsKeyFileEnvKey = "EDV_TLS_KEY_FILE"
 
 	dataVaultConfigurationStoreName = "data_vault_configurations"
+
+	sleep = 1 * time.Second
```
Comment (DRK3) [inline cmd/edv-rest/startcmd/start.go:99]: You can simplify this to just `time.Second`

File: cmd/edv-rest/startcmd/start.go [hunk 4]
```diff
@@ -94,14 +106,25 @@ var errInvalidDatabaseType = fmt.Errorf("database type not set to a valid type."
 
 var errCreateConfigStore = "failed to create data vault configuration store: %w"
 
+// nolint:gochecknoglobals
```
Comment (DRK3) [inline cmd/edv-rest/startcmd/start.go:109]: Reason for the `nolint`?

File: cmd/edv-rest/startcmd/start.go [hunk 5]
```diff
@@ -94,14 +106,25 @@ var errInvalidDatabaseType = fmt.Errorf("database type not set to a valid type."
...
+var supportedEDVStorageProviders = map[string]func(string, string) (edvprovider.EDVProvider, error){
+	databaseTypeCouchDBOption: func(databaseURL, prefix string) (edvprovider.EDVProvider, error) {
+		return couchdbedvprovider.NewProvider(databaseURL, prefix)
+	},
+	databaseTypeMemOption: func(_, _ string) (edvprovider.EDVProvider, error) { // nolint:unparam
```
Comment (DRK3) [inline cmd/edv-rest/startcmd/start.go:114]: Reason for the `nolint`?

File: .gitignore (hunk 1 above)
Comment (grace0925) [reply .gitignore:29]: it fixes codecov

File: cmd/edv-rest/startcmd/start.go (hunk 4 above)
Comment (DRK3) [reply cmd/edv-rest/startcmd/start.go:109]: What I meant was: is there an alternative approach that doesn't require global variables?

File: cmd/edv-rest/startcmd/start.go [hunk 6]
```diff
@@ -322,25 +335,55 @@ func setLogLevel(userLogLevel string) {
...
-		edvProv = couchDBEDVProv
-	default:
-		return edvProv, errInvalidDatabaseType
+	err := retry(func() error {
+		var openErr error
+		edvProv, openErr = providerFunc(parameters.databaseURL, parameters.databasePrefix)
+		return openErr
+	}, parameters.databaseTimeout)
+	if err != nil {
+		return nil, fmt.Errorf("fail to connect to %s: %w", parameters.databaseType, err)
```
Comment (DRK3) [inline cmd/edv-rest/startcmd/start.go:353]: Should be `failed` not `fail`

File: cmd/edv-rest/startcmd/start.go [hunk 7]
```diff
@@ -322,25 +335,55 @@ func setLogLevel(userLogLevel string) {
...
 	}
 
 	return edvProv, nil
 }
 
+// createConfigStore creates the config store, and creates indices if supported.
```
Comment (DRK3) [inline cmd/edv-rest/startcmd/start.go:359]: Shouldn't be a comma in this sentence

File: cmd/edv-rest/startcmd/start.go [hunk 8]
```diff
@@ -349,3 +392,12 @@ func constructCORSHandler(handler http.Handler) http.Handler {
 		},
 	).Handler(handler)
 }
+
+func retry(fn func() error, numRetries uint64) error {
+	return backoff.RetryNotify(fn,
+		backoff.WithMaxRetries(backoff.NewConstantBackOff(sleep), numRetries),
+		func(retryErr error, t time.Duration) {
+			logger.Warnf("failed to connect to database, will sleep for %s before trying again: %s\n",
```
Comment (DRK3) [inline cmd/edv-rest/startcmd/start.go:400]: The logger will add a newline automatically - you should remove the `\n`

Comment (DRK3) [reply cmd/edv-rest/startcmd/start.go:400]: (unless you really wanted two line breaks...?)

File: cmd/edv-rest/startcmd/start_test.go [hunk 9]
```diff
@@ -57,50 +85,52 @@ func TestStartCmdContents(t *testing.T) {
...
-	require.Equal(t,
-		"Neither host-url (command line flag) nor EDV_HOST_URL (environment variable) have been set.",
-		err.Error())
+		args := []string{"--" + hostURLFlagName, "localhost:8080", "--" + databaseTypeFlagName, "couchdb", "--" +
+			databaseTimeoutFlagName, "1"}
+		startCmd.SetArgs(args)
+
+		err := startCmd.Execute()
+
+		require.Equal(t, fmt.Errorf("fail to connect to couchdb: %w", couchdbedvprovider.ErrMissingDatabaseURL), err)
```
Comment (DRK3) [inline cmd/edv-rest/startcmd/start_test.go:107]: Should be `failed` not `fail`

File: cmd/edv-rest/startcmd/start_test.go [hunk 10]
```diff
@@ -236,18 +259,19 @@ func TestCreateProvider(t *testing.T) {
 		require.Equal(t, errInvalidDatabaseType, err)
 	})
 	t.Run("Error - CouchDB url is blank", func(t *testing.T) {
-		parameters := edvParameters{databaseType: databaseTypeCouchDBOption, databaseURL: ""}
+		parameters := edvParameters{databaseType: databaseTypeCouchDBOption, databaseURL: "", databaseTimeout: 1}
 
 		provider, err := createEDVProvider(&parameters)
 		require.Nil(t, provider)
-		require.Equal(t, couchdbedvprovider.ErrMissingDatabaseURL, err)
+		require.Equal(t, fmt.Errorf("fail to connect to couchdb: %w", couchdbedvprovider.ErrMissingDatabaseURL), err)
```
Comment (DRK3) [inline cmd/edv-rest/startcmd/start_test.go:266]: Should be `failed` not `fail`

This applies to other places too

File: cmd/edv-rest/startcmd/start_test.go [hunk 11]
```diff
@@ -259,34 +283,53 @@ func TestListenAndServe(t *testing.T) {
 }
 
 func TestActualDbInitializer_CreateConfigStore(t *testing.T) {
```
Comment (DRK3) [inline cmd/edv-rest/startcmd/start_test.go:285]: Now that the DB initializer is gone, this should be renamed

File: scripts/check_unit.sh [hunk 12]
```diff
@@ -10,11 +10,11 @@ echo "Running $0"
 
 go generate ./...
 pwd=`pwd`
-echo "" > "$pwd"/coverage.txt
+touch "$pwd"/coverage.out
 
 amend_coverage_file () {
 if [ -f profile.out ]; then
-     cat profile.out >> "$pwd"/coverage.txt
+     cat profile.out >> "$pwd"/coverage.out
```
Comment (DRK3) [inline scripts/check_unit.sh:17]: ?

File: cmd/edv-rest/startcmd/start.go (hunk 4 above)
Comment (DRK3) [reply cmd/edv-rest/startcmd/start.go:109]: Well, I guess it's ok if the only alternative is literally a method that just returns that value. Someone else can weigh in here

File: .gitignore (hunk 1 above)
Comment (DRK3) [reply .gitignore:29]: It's confusing to have an unrelated change like that in this PR. I suggest creating a new PR to just fix codecov, and get that merged in first

Comment (fqutishat) [reply .gitignore:29]: It's just one line

File: scripts/check_unit.sh (hunk 12 above)
Comment (grace0925) [reply scripts/check_unit.sh:17]: i'll separate this into a commit

File: .gitignore (hunk 1 above)
Comment (DRK3) [reply .gitignore:29]: @fqutishat I also see changes to the `check_unit.sh` script too... so not just one line...

At the very least, adding "also fixes codecov" or something similar to the commit body message will help future people understand why those lines were changed.

Comment (DRK3) [reply .gitignore:29]: Having the separate commit also makes it easier to figure out what fixed what. e.g. when you're in the IDE and you use the annotate feature, it'll show that line changed and when you click on it, you can see a commit that describes exactly what happened "fix: codecov". With the change in this commit, you see something confusing: "feat: configure database connection retries"

Comment (DRK3) [reply .gitignore:29]: Adding a "also fixes codecov" to the commit body message is good enough for me

Comment (grace0925) [reply .gitignore]: fixed

File: cmd/edv-rest/startcmd/start_test.go (hunk 11 above)
Comment (grace0925) [reply cmd/edv-rest/startcmd/start_test.go]: fixed

File: cmd/edv-rest/startcmd/start_test.go (hunk 10 above)
Comment (grace0925) [reply cmd/edv-rest/startcmd/start_test.go]: fixed

File: cmd/edv-rest/startcmd/start_test.go (hunk 9 above)
Comment (grace0925) [reply cmd/edv-rest/startcmd/start_test.go]: fixed

File: cmd/edv-rest/startcmd/start.go (hunk 8 above)
Comment (grace0925) [reply cmd/edv-rest/startcmd/start.go]: fixed

File: cmd/edv-rest/startcmd/start.go (hunk 7 above)
Comment (grace0925) [reply cmd/edv-rest/startcmd/start.go]: fixed

File: cmd/edv-rest/startcmd/start.go (hunk 6 above)
Comment (grace0925) [reply cmd/edv-rest/startcmd/start.go]: fixed

File: cmd/edv-rest/startcmd/start.go (hunk 3 above)
Comment (grace0925) [reply cmd/edv-rest/startcmd/start.go]: fixed

File: cmd/edv-rest/startcmd/start.go (hunk 2 above)
Comment (grace0925) [reply cmd/edv-rest/startcmd/start.go]: fixed

**Review by rolsonquadras — APPROVED** [review]

**Review by sudeshrshetty — APPROVED** [review]

**Review by DRK3 — APPROVED** [review]

**Pull request merged**